include/envoy/event/dispatcher.h

```diff
@@ -30,8 +30,8 @@ namespace Event {
  */
 // clang-format off
 #define ALL_DISPATCHER_STATS(HISTOGRAM)                                                            \
-  HISTOGRAM(loop_duration_us)                                                                      \
-  HISTOGRAM(poll_delay_us)
+  HISTOGRAM(loop_duration, Microseconds)                                                           \
```

Comment:
I see how this is a positive change. However I wonder if this might be a breaking change for monitoring setups for some deployments. How will we message this out and manage this transition?

Reply:
I've been thinking about this change before making it and came to a conclusion that however nasty this may be, the alternative is no better (potentially worse).

This is a breaking change as people definitely have alerts set up for these. If left as-is, `statsd` change is not possible (automatically append the actually correct symbol) and Metric sinks supporting generic histograms vs timers will report milli-milliseconds, micro-milliseconds and other nonsense (they would take values in seconds and scale using metric prefixes and display the scaled value so the metric name suffix will mismatch). It's also simply put messy. Some of the reported histograms right now are rather confusing already (Redis command latency is reported in millis or micros depending on a config value) so alerting on these is error-prone at best.

Having said that we may create a table of old -> new metric names for `statsd` and for `non-statsd` sinks so people can do a low effort find & replace instead of figuring it out? This can be published with release itself as a pre-read.

Thoughts?

Reply:
@htuch @alyssawilk should this be considered an API change and versioned as such?

Reply:
To sanity check, this is changing the variable names for all statsd variables?  I definitely think that's something that folks should be able to migrate to carefully and not have to fix as they pull HEAD.
Ideally it's the sort of thing we could "runtime" guard and do away with but I don't think it's the sort of thing you can flip live.  Given that I'd lean towards a config option which we deprecate to (eventually) avoid code cruft

Reply:
@alyssawilk this affects histogram names, I've compiled the changes to make this easier to understand:


### Histogram name changes for non-statsd sinks (unit symbol suffix dropped if present)

| Group       |                        Old | New                    |
| :---------- | -------------------------: | :--------------------- |
| Cluster     |    `upstream_cx_length_ms` | `upstream_cx_length`   |
| Cluster     |   `upstream_cx_connect_ms` | `upstream_cx_connect`  |
| Dispatcher  |            `poll_delay_us` | `poll_delay`           |
| Dispatcher  |         `loop_duration_us` | `loop_duration`        |
| Dubbo       |          `request_time_ms` | `request_time`         |
| Dynamo      |         `upstream_rq_time` | same                   |
| Dynamo      |   `upstream_rq_time_{i}xx` | same                   |
| Dynamo      | `upstream_rq_time_unknown` | same                   |
| HTTPConnMan |       `downstream_rq_time` | same                   |
| HTTPConnMan |  `downstream_cx_length_ms` | `downstream_cx_length` |
| Listener    |  `downstream_cx_length_ms` | `downstream_cx_length` |
| Mongo       |                  `time_ms` | `time`                 |
| Mongo       |               `reply_size` | same                   |
| Mongo       |            `reply_time_ms` | `reply_time`           |
| Redis       |                  `latency` | same                   |
| Redis       |         `upstream_rq_time` | same                   |
| Server      |   `initialization_time_ms` | `initialization_time`  |
| Thrift      |          `request_time_ms` | `request_time`         |
| UserAgent   |  `downstream_cx_length_ms` | `downstream_cx_length` |
| ZK          |         `<opcode>_latency` | same                   |
| ZK          | `connect_response_latency` | same                   |

### Histogram name changes for statsd sinks (unit symbol suffix appended where absent)

| Group       |                        Old | New                                         |
| :---------- | -------------------------: | :------------------------------------------ |
| Cluster     |    `upstream_cx_length_ms` | same                                        |
| Cluster     |   `upstream_cx_connect_ms` | same                                        |
| Dispatcher  |            `poll_delay_us` | same                                        |
| Dispatcher  |         `loop_duration_us` | same                                        |
| Dubbo       |          `request_time_ms` | same                                        |
| Dynamo      |         `upstream_rq_time` | `upstream_rq_time_ms`                       |
| Dynamo      |   `upstream_rq_time_{i}xx` | `upstream_rq_time_{i}xx_ms`                 |
| Dynamo      | `upstream_rq_time_unknown` | `upstream_rq_time_unknown_ms`               |
| HTTPConnMan |       `downstream_rq_time` | `downstream_rq_time_ms`                     |
| HTTPConnMan |  `downstream_cx_length_ms` | same                                        |
| Listener    |  `downstream_cx_length_ms` | same                                        |
| Mongo       |                  `time_ms` | same                                        |
| Mongo       |               `reply_size` | `reply_size_b`                              |
| Mongo       |            `reply_time_ms` | same                                        |
| Redis       |                  `latency` | `latency_us` or `latency_ms` (configurable) |
| Redis       |         `upstream_rq_time` | `upstream_rq_time_ms`                       |
| Server      |   `initialization_time_ms` | same                                        |
| Thrift      |          `request_time_ms` | same                                        |
| UserAgent   |  `downstream_cx_length_ms` | same                                        |
| ZK          |         `<opcode>_latency` | `<opcode>_latency_ms`                       |
| ZK          | `connect_response_latency` | `connect_response_latency_ms`               |

Reply:
IMO it's a non-starter to change the stat names. This is too disruptive to existing users.

I haven't yet reviewed this PR but can we please not change any stat names?

Reply:
Sure, I can revert the name changes and the `statsd` suffixing so all histogram names are always reported as-is. Is this going to be the recommendation for new histograms as well?

Right now Envoy own histograms are all suffixed (except 1 `downstream_rq_time` in `http_conn_man`) but extensions are all over the place. If you decide to always take the name as-is, we can maybe add a formatter check for name suffixes not matching the provided unit?

Reply:
Sure, I think it's great to require good name hygiene for new stats, I just don't think it's worth the end-user pain to change the names of the existing ones (though I agree it would be nice).

Reply:
ok so to recap:
1) Keep all existing names as-is.
2) Do not autosuffix in `statsd`.
3) Have a `check_format` rule ensuring no new histograms are suffixed but whitelist all existing.

I will address the remaining comments and send 1 commit to undo the renames so it's easier to review.

Reply:
I will leave the `statsd` to the sink maintainers, but recommendation is that since `statsd` assumes all histograms are millis, all other time values should be scaled to millis on flush unless they are suffixed. This is the minimally invasive fix to reinstall sanity. This way things like Redis `latency` would be scaled to millis and things like `poll_delay_us` would be reported as-is but it would be clear on the UI it's micros (due to the name). And for all new stats they can be assumed to be millis, no suffixed, scaled to millis if necessary.

Reply:
A little late to the discussion, but I agree that this is too disruptive. However, stats are not (today) part of the Envoy API, so they don't live by its versioning guidelines. Is there a way we can add versioning to the stats interface to allow it to evolve over time? Should we provide a canonical proto definition that will allow it to move alongside the API major versions, or perhaps on its own clock?

Reply:
I overall agree this is a breaking change and probably can't be released just like this but *I'd really welcome any strategy for dealing with metric name migrations* as little things here and there keep being added over time and at some point consistency and clarity is lost in the name of backward compatibility and at that point it's hard to say what causes more bugs - vague metrics and wrong alerts set up based on misinterpretation or broken alerts due to the name changes. It's not a pressing problem but would be could if there was a solution at some point.